src/hotspot/share/gc/shared/slidingForwarding.cpp

```diff
@@ -0,0 +1,150 @@
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+#include "gc/shared/gc_globals.hpp"
+#include "gc/shared/slidingForwarding.hpp"
+#include "utilities/ostream.hpp"
+#include "utilities/powerOfTwo.hpp"
+
+#ifdef _LP64
+
+HeapWord* const SlidingForwarding::UNUSED_BASE = reinterpret_cast<HeapWord*>(0x1);
+
+SlidingForwarding::SlidingForwarding(MemRegion heap, size_t region_size_words)
+  : _heap_start(heap.start()),
+    _num_regions(((heap.end() - heap.start()) / region_size_words) + 1),
+    _region_size_words_shift(log2i_exact(region_size_words)),
+  _target_base_table(nullptr),
+  _fallback_table(nullptr) {
+  assert(_region_size_words_shift <= NUM_COMPRESSED_BITS, "regions must not be larger than maximum addressing bits allow");
+  size_t heap_size_words = heap.end() - heap.start();
+  if (UseSerialGC && heap_size_words <= (1 << NUM_COMPRESSED_BITS)) {
+    // In this case we can treat the whole heap as a single region and
+    // make the encoding very simple.
+    _num_regions = 1;
+    _region_size_words_shift = log2i_exact(round_up_power_of_2(heap_size_words));
+  }
+}
+
+SlidingForwarding::~SlidingForwarding() {
+  if (_target_base_table != nullptr) {
+    FREE_C_HEAP_ARRAY(HeapWord*, _target_base_table);
+  }
+  if (_fallback_table != nullptr) {
+    delete _fallback_table;
+  }
+}
+
+void SlidingForwarding::begin() {
+  assert(_target_base_table == nullptr, "Should be uninitialized");
+  _target_base_table = NEW_C_HEAP_ARRAY(HeapWord*, _num_regions * NUM_TARGET_REGIONS, mtGC);
+  size_t max = _num_regions * NUM_TARGET_REGIONS;
+  for (size_t i = 0; i < max; i++) {
+    _target_base_table[i] = UNUSED_BASE;
+  }
+}
+
+void SlidingForwarding::end() {
+  assert(_target_base_table != nullptr, "Should be initialized");
+  FREE_C_HEAP_ARRAY(HeapWord*, _target_base_table);
+  _target_base_table = nullptr;
+
+  if (_fallback_table != nullptr) {
+    delete _fallback_table;
+    _fallback_table = nullptr;
+  }
+}
+
+void SlidingForwarding::fallback_forward_to(HeapWord* from, HeapWord* to) {
+  if (_fallback_table == nullptr) {
+    _fallback_table = new FallbackTable();
+  }
+  _fallback_table->forward_to(from, to);
+}
+
+HeapWord* SlidingForwarding::fallback_forwardee(HeapWord* from) const {
+  if (_fallback_table == nullptr) {
+    return nullptr;
```

Comment:
Is it an error to ask for `fallback_forwardee` if there is no table initialized? This implies that no fallback forwardings happened, which implies the no fallback mask is set in any object header, so why are we at this path?